### PR TITLE
Document how to reindex a TSDS

### LINKED ISSUE #99176: Document how to reindex a tsdb data stream
Re-indexing a tsdb data stream is more challenging than re-indexing a regular tsdb data stream. This is because when a new data stream is created, then the new backing index start and end time settings are blindly set to $now-2h to $now+2h. The backing indices of the existing tsdb data stream may overlap with this, or not at all. Directly re-indexing from the old tsdb data stream into the new tsdb data stream only works for documents that have a timestamp that matches with $now-2h to $now+2h.

Given that reindexing a tsdb data stream is possible, we should document how to do this. Currently no documentation around reindex a tsdb data stream exists. 

The process looks something like this:
- Create a specific index template for the new data stream only that will contain the re-indexed data. Otherwise other data streams may get affected. This index template should contain the new mappings / index settings that should get applied.
- Update the template to set specific `index.time_series.start_time` and `index.time_series.end_time` index settings. The start and end time settings should be based on the lowest and highest `@timestamp` values in the data stream to be reindex. This way the first backing index is fixed to contain all data that is contained in the data stream that should be reindexed.
- Update the template to set the `index.number_of_shards` index setting to the sum of all primary shards of all backing indices of the data stream to be reindexed.
- Update the template to set `index.number_of_replicas` to zero and unset the `index.lifecycle.name` index setting.
- Start the reindex operation.
- After reindexing completed then remove the `index.time_series.start_time`, `index.time_series.end_time` index settings from the template and set `index.number_of_replicas`, `index.number_of_shards` and `index.lifecycle.name` to the original values.
- Invoke the rollover api without any conditions set. Now data stream should be ready accept recent data.

### PR DESCRIPTION
**Pull request by kkrik-es**
Time-series data streams require updating start and end times in the destination index template, to avoid errors during copying of older docs.

Fixes #99176

### CODE REVIEW

**Comment by kkrik-es** [comment]
@elasticsearchmachine  run elasticsearch-ci/docs

**Comment by kkrik-es** [comment]
@elasticsearchmachine run elasticsearch-ci/docs

File: docs/reference/data-streams/tsds-reindex.asciidoc [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+[source,console]
+----
+POST /k9s/_rollover/
+----
+// TEST[continued]
+
+This creates a new backing index with the updated index settings. The destination data stream is now ready to accept new docs.
+
+Note that the initial backing index can still accept docs within the range of timestamps derived from the source data
```
Comment (martijnvg) [inline docs/reference/data-streams/tsds-reindex.asciidoc:291]: Maybe also add a note about data streams containing downsampled indices?

**Review by martijnvg — APPROVED** [review]
LGTM

File: docs/reference/data-streams/tsds-reindex.asciidoc (hunk 1 above)
Comment (kkrik-es) [reply docs/reference/data-streams/tsds-reindex.asciidoc:291]: I added one at the top, you want another one here? No objection, just checking..

Comment (martijnvg) [reply docs/reference/data-streams/tsds-reindex.asciidoc:291]: oops missed that... never mind

File: docs/reference/data-streams/tsds-reindex.asciidoc [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+----
+// NOTCONSOLE
+
+To reindex this TSDS, do not to re-use its index template in the destination data stream, to avoid impacting its
+functionality. Instead, clone the template of the source TSDS and apply the following modifications:
+
+* Set `index.time_series.start_time` and `index.time_series.end_time` index settings explicitly. Their values should be
+based on the lowest and highest @timestamp values in the data stream to reindex. This way, the initial backing index can
+load all data that is contained in the source data stream.
+* Set `index.number_of_shards index` setting to the sum of all primary shards of all backing indices of the source data
```
Comment (abdonpijpelink) [inline docs/reference/data-streams/tsds-reindex.asciidoc:157]:
```suggestion
* Set `index.number_of_shards` index setting to the sum of all primary shards of all backing indices of the source data
```

File: docs/reference/data-streams/tsds-reindex.asciidoc [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+// TEST[continued]
+
+[discrete]
+[[tsds-reindex-restore]]
+==== Restore the destination index template
+
+Once the reindexing operation completes, restore the index template for the destination TSDS as follows:
+
+* Remove the overrides for `index.time_series.start_time` and `index.time_series.end_time`.
+* Restore the values of `index.number_of_shards index`, `index.number_of_replicas`  and  `index.lifecycle.name` as
```
Comment (abdonpijpelink) [inline docs/reference/data-streams/tsds-reindex.asciidoc:243]:
```suggestion
* Restore the values of `index.number_of_shards`, `index.number_of_replicas`  and  `index.lifecycle.name` as
```

File: docs/reference/data-streams/tsds-reindex.asciidoc (hunk 1 above)
Comment (abdonpijpelink) [inline docs/reference/data-streams/tsds-reindex.asciidoc:291]:
```suggestion
This creates a new backing index with the updated index settings. The destination data stream is now ready to accept new documents.

Note that the initial backing index can still accept documents within the range of timestamps derived from the source data
```

**Review by abdonpijpelink — COMMENTED** [review]
I've left a few minor suggestions.

**Comment by kkrik-es** [comment]
> I've left a few minor suggestions.

Thanks, good catch with `end_time` - copy/paste added insult to injury..

**Comment by kkrik-es** [comment]
@elasticsearchmachine run elasticsearch-ci/docs

**Review by abdonpijpelink — APPROVED** [review]

**Comment by agithomas** [comment]
@kkrik-es , will this change create a new link in the below mentioned navbar / menu in v8.11 ? 

We plan to reference the documentation from the integration package Readme.  Please advice. 



image

**Comment by kkrik-es** [comment]
> @kkrik-es , will this change create a new link in the below mentioned navbar / menu in v8.11 ?
> 
> We plan to reference the documentation from the integration package Readme. Please advice.
> 
> image

This change actually adds the following link to the navbar:

Screenshot 2023-10-17 at 11 05 51 AM

**Pull request merged**